ampd/src/event_processor.rs

```diff
@@ -79,24 +55,87 @@ impl EventProcessor {
 
     pub fn add_handler<H, S, E>(&mut self, handler: H, event_stream: S) -> &mut Self
     where
-        H: EventHandler + Send + Sync + 'static,
+        H: EventHandler + Send + 'static,
         S: Stream<Item = Result<Event, E>> + Send + 'static,
         E: Context,
     {
-        self.tasks
-            .push(consume_events(event_stream, handler, self.token.child_token()).into());
+        self.tasks.push(Box::pin(EventProcessor::consume_events(
+            event_stream,
+            handler,
+            self.token.child_token(),
+        )));
         self
     }
 
-    pub async fn run(self) -> Result<(), EventProcessorError> {
-        let handles = self.tasks.into_iter().map(tokio::spawn);
+    pub async fn run(self) -> Result<(), Error> {
+        let mut join_set = JoinSet::new();
+
+        for task in self.tasks.into_iter() {
+            // tasks clean up on their own after the cancellation token is triggered, so we discard the abort handles
+            join_set.spawn(task);
+        }
+
+        EventProcessor::wait_for_completion(&mut join_set, &self.token).await
+    }
+
+    async fn consume_events<H, S, E>(
+        event_stream: S,
+        handler: H,
+        token: CancellationToken,
+    ) -> Result<(), Error>
+    where
+        H: EventHandler,
+        S: Stream<Item = Result<Event, E>>,
+        E: Context,
+    {
+        let mut event_stream = Box::pin(event_stream);
+        while let Some(event) = event_stream.next().await {
+            let event = event.change_context(Error::EventStream)?;
+
+            handler
+                .handle(&event)
+                .await
+                .change_context(Error::Handler)?;
+
+            if matches!(event, Event::BlockEnd(_)) && token.is_cancelled() {
+                break;
+            }
+        }
 
-        try_join_all(handles)
-            .await
-            .change_context(EventProcessorError::EventHandlerError)?
-            .into_iter()
-            .find(Result::is_err)
-            .unwrap_or(Ok(()))
+        Ok(())
+    }
+
+    async fn wait_for_completion(
+        join_set: &mut JoinSet<Result<(), Error>>,
+        token: &CancellationToken,
+    ) -> Result<(), Error> {
+        let mut extended_error = None;
+        while let Some(result) = join_set.join_next().await {
+            token.cancel();
+
+            extended_error = match result.change_context(Error::Tasks) {
+                Err(err) | Ok(Err(err)) => EventProcessor::extend_err(extended_error, err),
+                Ok(_) => extended_error,
+            };
+        }
+
+        if let Some(error) = extended_error {
+            Err(error)
+        } else {
+            Ok(())
+        }
+    }
+
+    fn extend_err(
```

Comment:
I don't get this. Why cancel the token in a loop for each of the exiting tasks?

Reply:
it's just easier this way, any further call to cancel it after the first one is a noop, so I don't have to add a guard clause. However, I can add a comment explaining this